### PR TITLE
Separate Excel Reporter from HTML Reporter artifacts.

### PR DESCRIPTION
**Pull request by squash-merge**
Following content are added in a new worksheet in the ExcelReport:
1. Links to the page source.
2. Links to the screen capture of the page.
3. Parameters to the test method.
4. Other TestNG output logs.
5. The stacktrace of all the failures.

Also changes to disable HTML Reporter.

### CODE REVIEW

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java [hunk 1]
```diff
@@ -110,6 +117,20 @@ public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String
         }
     }
 
+    /*
+     * Set the output file name for the Excel Report. The file should have 'xls' extension. *
+     * 
+     * @param fileName The file name without any path.
+     */
+    public void setExcelFileName(String fileName) {
```
Comment (krmahadevan) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java:130]: Minor observation : We could have leveraged `Preconditions.checkArgument`

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java [hunk 2]
```diff
@@ -91,18 +95,70 @@ public void setITestResultobj(ITestResult iTestResult) {
...
         }
 
-        this.sSSLink = "file:" + File.separator + File.separator + sOutputDir + File.separator + "html"
-                + File.separator + singleMethod.getId() + ".html";
+        this.sSSMsg = makeContentForLinksColumn(iTestResult);
 
         // if failed, then add error details
         if (iStatus == ITestResult.FAILURE) {
-            lstError.add(iTestResult.getThrowable().toString());
+            String stacktrace = StringUtils.defaultString(StringUtils.substringBetween(
```
Comment (krmahadevan) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java]: I would suggest that you please break this down into individual calls. It becomes difficult to read and also difficult to debug as well.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                String[] outputLines = output.split(":", 2);
+                String outputColumn = outputLines[0];
+                String outputValue = (outputLines.length > 1) ? outputLines[1].trim() : output.trim();
+
+                int colNum = iStartColNum;
+                row = sheet.createRow(rowNum);
+                col = row.createCell(colNum);
+
+                col.setCellStyle(Styles.getSubHeading2StyleThinBorder());
+                if (outputLines.length > 1) {
```
Comment (krmahadevan) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java]: Can be changed to

```
String cellValue="";
if (outputLines.length > 1) {
     cellValue=outputColumn;
}
col.setCellValue(cellValue);
```

File: client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java [hunk 4]
```diff
@@ -59,6 +84,34 @@ public void testExcelReporter() {
...
+                fileInputStream = new FileInputStream(p.toFile());
+                workbook = new HSSFWorkbook(fileInputStream);
+                assertEquals(ReportSheetNames.values().length, workbook.getNumberOfSheets(),
+                        "Verify number of worksheets in ExcelReport");
+            } catch (FileNotFoundException e) {
+                fail("Test excel-report-test failed due to FileNotFoundException.");
+            } catch (IOException e) {
+                fail("Test excel-report-test failed due to IOException.");
+            } finally {
+                if (fileInputStream != null) {
```
Comment (krmahadevan) [inline client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java]: Please use `IOUtils.closeQuietyly(fileInputStream);`

File: client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java [hunk 5]
```diff
@@ -59,6 +84,34 @@ public void testExcelReporter() {
...
+
+        if (Files.exists(p) && Files.isRegularFile(p)) {
+            FileInputStream fileInputStream = null;
+            HSSFWorkbook workbook = null;
+            try {
+                fileInputStream = new FileInputStream(p.toFile());
+                workbook = new HSSFWorkbook(fileInputStream);
+                assertEquals(ReportSheetNames.values().length, workbook.getNumberOfSheets(),
+                        "Verify number of worksheets in ExcelReport");
+            } catch (FileNotFoundException e) {
```
Comment (krmahadevan) [inline client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java]: I dont think we resort to this practice of catching exceptions and explicitly failing the tests. By default TestNG fails a test if it throws an exception. So we could perhaps clear up this part.

File: client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java [hunk 6]
```diff
@@ -35,6 +47,19 @@
 public class ExcelReportTest {
 
     private static final String SUREFIRE_DEFAULT_REPORTS_DIR = "target/surefire-reports";
+    private String excelReportFileName = "dummy_excel_test.xls";
+    private Path p = Paths.get(SUREFIRE_DEFAULT_REPORTS_DIR, excelReportFileName);
+
+    @BeforeMethod(groups = { "excel-report-test" })
+    public void removeExcelFileBeforeTest() {
```
Comment (krmahadevan) [inline client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java]: We could change this to 

```
        if (Files.isRegularFile(p)) {
            FileUtils.deleteQuietly(p.toFile());
        }

```

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java [hunk 7]
```diff
@@ -48,8 +66,18 @@ public boolean isEnabled() {
...
     static boolean getBooleanValFromVMArg(String vmArgValue) {
-        boolean flag = true;
+        return getBooleanValFromVMArg(vmArgValue, defaultIsEnabled);
+    }
+
+    /*
+     * Returns boolean value of the JVM argument when available, else returns the {@code isEnabledWhenJvmAbsent}.
+     */
+    static boolean getBooleanValFromVMArg(String vmArgValue, boolean isEnabledWhenJvmAbsent) {
+        boolean flag = isEnabledWhenJvmAbsent;
```
Comment (krmahadevan) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: This variable is being assigned but I dont see it being used.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java (hunk 1 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java:130]: Addressed: Used `Preconditions.checkArgument`.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java (hunk 2 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java]: Addressed: Split the statement into two. First may return null, which is handled in the second.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java (hunk 3 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java]: Addressed: Moved the condition to declaration of `String outputColumn`.

File: client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java (hunk 4 above)
Comment (squash-merge) [inline client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java]: Addressed: Used `IOUtils.closeQuietyly(fileInputStream)` as suggested.

File: client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java (hunk 5 above)
Comment (squash-merge) [inline client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java]: Addressed: Reduced error handling and removed explicit fail calls.

File: client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java (hunk 6 above)
Comment (squash-merge) [inline client/src/test/java/com/paypal/selion/reports/reporter/excelreport/ExcelReportTest.java]: Addressed: Conditionally used `FileUtils.deleteQuietly` as suggested.

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java (hunk 7 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: @krmahadevan, I did not get this comment. I have renamed member of `class ListenerInfo` and the method parameters to make its readability better.

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java [hunk 8]
```diff
@@ -33,24 +38,62 @@
...
-        boolean flag = true;
+        return getBooleanValFromVMArg(vmArgValue, ENABLE_LISTENER_BY_DEFAULT);
+    }
+
+    /**
+     * Returns boolean value of the JVM argument when available, else returns the {@code isEnabledWhenJvmAbsent}.
+     * 
+     * @param vmArgValue
+     *      The VM argument name.
+     * @param isListenerEnabledWhenJvmAbsent
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: How can the JVM be absent in a java process? I suspect you mean to call this more like `defaultStateWhenNotDefined`

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java [hunk 9]
```diff
@@ -33,24 +38,62 @@
...
+     * @param vmArgValue
+     *          
+     */
     static boolean getBooleanValFromVMArg(String vmArgValue) {
-        boolean flag = true;
+        return getBooleanValFromVMArg(vmArgValue, ENABLE_LISTENER_BY_DEFAULT);
+    }
+
+    /**
+     * Returns boolean value of the JVM argument when available, else returns the {@code isEnabledWhenJvmAbsent}.
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: the `{@code` portion of this javadoc is incorrect

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java [hunk 10]
```diff
@@ -33,24 +38,62 @@
      *            - The JVM argument that should be read to decide if the listener is to be enabled/disabled.
      */
     public ListenerInfo(Class<?> className, String jvmArgToParse) {
+        this(className, jvmArgToParse, ENABLE_LISTENER_BY_DEFAULT);
+    }
+
+    /**
+     * Saves information about listener class which is enabled/disabled per its VM argument. When VM argument is
+     * unavailable, it's enabled/disabled is decided by the caller via (@code isEnabledWhenJvmAbsent}.
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: the {@code portion of this javadoc is incorrect

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java [hunk 11]
```diff
@@ -33,24 +38,62 @@
...
+     * 
+     * @param className
+     *            - The Class name.
+     * @param jvmArgToParse
+     *            - The JVM argument that should be read to decide if the listener is to be enabled/disabled.
+     * @param isListenerEnabledWhenJvmAbsent
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: How can the JVM be absent in a java process? I suspect you mean to call this more like `defaultStateWhenNotDefined`

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java [hunk 12]
```diff
@@ -402,32 +434,42 @@ private void generateTCBasedData(ArrayList<TestCaseResult> allTestsList) {
             str.add(tcResult.getTestDesc());
             str.add(tcResult.getGroup().toString());
             str.add(String.valueOf(tcResult.getDurationTaken()));
-            str.add(tcResult.getsslink());
+//            str.add(tcResult.getsslink());
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java]: remove the commented code

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java [hunk 13]
```diff
@@ -19,6 +19,9 @@
 import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.nio.file.Files;
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java:21]: the import `java.io.File` is no longer used in this class

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java [hunk 14]
```diff
@@ -30,6 +33,7 @@
 import java.util.Map.Entry;
 import java.util.logging.Level;
 
+import org.apache.commons.lang3.StringUtils;
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java]: let's stick with `org.apache.commons.lang.StringUtils` unless there is something specific you need in `lang3` here

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java [hunk 15]
```diff
@@ -87,9 +87,13 @@ public void addToGeneratedReport(List<BaseReport<V>> lstCustomReport) {
         for (String indReport : this.reportData.keySet()) {
             if (iTypeOfReport == 0) {
                 b = new SummaryReport(indReport);
-            } else {
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java:90]: the previous way of formatting the `else` conditions is preferred.  Use `} else {` on the same line.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java [hunk 16]
```diff
@@ -87,9 +87,13 @@ public void addToGeneratedReport(List<BaseReport<V>> lstCustomReport) {
         for (String indReport : this.reportData.keySet()) {
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java:88]: Please parameterize `BaseReport` on line 86

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+|  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for  |
+|  the specific language governing permissions and limitations under the License.                                     |
+\*-------------------------------------------------------------------------------------------------------------------*/
+
+package com.paypal.selion.reports.reporter.excelreport;
+
+import java.util.List;
+
+import org.apache.commons.collections.ListUtils;
+import org.apache.commons.lang.StringUtils;
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java]: this import is not used.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/BaseReport.java [hunk 18]
```diff
@@ -21,6 +21,7 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.commons.lang.StringUtils;
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/BaseReport.java]: this import is not used.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java [hunk 19]
```diff
@@ -232,8 +255,16 @@ private void createReportMap() {
         BaseReport<List<String>> bR = (BaseReport<List<String>>) fullReportMap.get(fullReportMap.size() - 1)
                 .getGeneratedReport().iterator().next();
         List<String> lsTitles = Arrays.asList(new String[] { "Class Name", "Method/Testcase id", "Test Description",
-                "Group[s]", "Time taken", "Link", "Error Message", "Error Details" });
+                "Group[s]", "Time taken", "Output Logs", "Error Message", "Error Details" });
         bR.setColTitles(lsTitles);
+        
+        // TestCase Output Details Report
+        Map<String, List<List<String>>> fifthTestOutputSubReportMap = new LinkedHashMap<String, List<List<String>>>();
+        fifthTestOutputSubReportMap.put("Tests Output", tcOutputData);
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java]: Lets' call this tab `Test Output` instead.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java [hunk 20]
```diff
@@ -40,11 +45,11 @@
     private String testDescription;
     private List<String> sGroup;
     private long durationTaken;
-    private ArrayList<String> sSSMsg;
-    private String sSSLink;
+    private List<String> sSSMsg;
     private int iStatus;
 
     public static final String NA = "NA";
+    public static final String newLine = "\n";
```
Comment (mach6) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java]: Sonar will complain here.. Constants should have all upppercase names

Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestCaseResult.java]: Addressed: Changed to uppercase name.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java (hunk 19 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java]: Addressed: This is the section header. Changed this to 'Test Output'. Will also update ReportSheetNames#TESTOUTPUTDETAILSREPORT to use 'Test Output'.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/BaseReport.java (hunk 18 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/BaseReport.java]: Addressed: Removed the import.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java (hunk 17 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/TestOutputReport.java]: Addressed: Removed unused import.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java (hunk 16 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java:88]: Addressed: Added a suppress for raw types.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java (hunk 15 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ReportMap.java:90]: Addressed: Changed to `} else {`.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java (hunk 14 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java]: Addressed: Changed lang3 imports to lang for client code in another commit.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java (hunk 13 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java:21]: Addressed: Removed unused import.

File: client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java (hunk 12 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/reports/reporter/excelreport/ExcelReport.java]: addressed: removed comment.

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java (hunk 11 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: Addressed: Used conventional names and comments.

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java (hunk 10 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: Addressed: Updated the **@{code** block.

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java (hunk 9 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: Addressed: Updated the @{code block.

File: client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java (hunk 8 above)
Comment (squash-merge) [inline client/src/main/java/com/paypal/selion/configuration/ListenerInfo.java]: Addressed: Used conventional names and comments.

**Comment by mach6** [comment]
this PR has been updated and merged.

**Pull request closed without merging**